tests/loggen/socket_plugin/socket_plugin.c

```diff
@@ -0,0 +1,409 @@
+  g_mutex_unlock(thread_lock);
+
+  close(fd);
+  g_thread_exit(NULL);
+  return NULL;
+}
+
+gpointer
+active_thread_func(gpointer user_data)
+{
+  thread_data *thread_context = (thread_data *)user_data;
+  plugin_option *option = thread_context->option;
+
+  int sock_type = SOCK_STREAM;
+
+  if (sock_type_d)
+    sock_type = SOCK_DGRAM;
+  if (sock_type_s)
+    sock_type = SOCK_STREAM;
+
+  char *message = g_malloc0(MAX_MESSAGE_LENGTH+1);
+
+  int fd;
+  if (unix_socket_x)
+    fd = connect_unix_domain_socket(sock_type, option->target);
+  else
+    fd = connect_ip_socket(sock_type, option->target, option->port, option->use_ipv6);
+
+  if (fd<0)
+    {
+      ERROR("can not connect to %s:%s (%p)\n",option->target, option->port,g_thread_self());
+    }
+  else
+    {
+      DEBUG("(%d) connected to server on socket %d (%p)\n",thread_context->index,fd,g_thread_self());
+    }
+
+  g_mutex_lock(thread_lock);
+  connect_finished++;
+  g_mutex_unlock(thread_lock);
+
+  if (connect_finished == active_thread_count + idle_thread_count)
+    g_cond_broadcast(thread_connected);
+
+  DEBUG("thread (%s,%p) created. wait for start ...\n",loggen_plugin_info.name,g_thread_self());
+  g_mutex_lock(thread_lock);
+  while (!thread_run)
+    {
+      g_cond_wait(thread_start,thread_lock);
+    }
+  g_mutex_unlock(thread_lock);
+
+  DEBUG("thread (%s,%p) started. (r=%d,c=%d)\n",loggen_plugin_info.name,g_thread_self(),option->rate,
+        option->number_of_messages);
+
+  unsigned long count = 0;
+  thread_context->buckets = thread_context->option->rate - (thread_context->option->rate / 10);
+
+  gettimeofday(&thread_context->last_throttle_check, NULL);
+  gettimeofday(&thread_context->start_time, NULL);
+
+  gboolean connection_error = FALSE;
+
+  while (fd>0 && thread_run && !connection_error)
+    {
+      if (thread_check_exit_criteria(thread_context))
+        break;
+
+      if (thread_check_time_bucket(thread_context))
+        continue;
+
+      if (!loggen_plugin_info.generate_message)
+        {
+          ERROR("generate_message not yet set up(%p)\n",g_thread_self());
+          break;
+        }
+
+      int str_len = loggen_plugin_info.generate_message(message, MAX_MESSAGE_LENGTH, thread_context->index, count++);
+
+      if (str_len < 0)
+        {
+          ERROR("can't generate more log lines. end of input file?\n");
+          break;
+        }
+
+      ssize_t sent = 0;
+      while (sent < strlen(message))
+        {
+          ssize_t rc = send_plain(fd, message + sent, strlen(message) - sent);
+          if (rc < 0)
+            {
+              ERROR("error sending buffer on %d (rc=%zd)\n",fd,rc);
+              errno = ECONNABORTED;
+              connection_error = TRUE;
+              break;
+            }
+          sent += rc;
+        }
+
+      thread_context->sent_messages++;
+      thread_context->buckets--;
+    }
+  DEBUG("thread (%s,%p) finished\n",loggen_plugin_info.name,g_thread_self());
+
+  free(message);
```

Comment:
Let's use g_free for a g_malloc0-ed area.

Reply:
done